components/infusionsoft/sources/common.ts

```diff
@@ -0,0 +1,93 @@
+import infusionsoft from "../app/infusionsoft.app";
+import { SourceHttpRunOptions } from "@pipedream/types";
+import { createHookParams } from "../types/requestParams";
+import { webhook, webhookObject } from "../types/responseSchemas";
+
+export default {
+  props: {
+    infusionsoft,
+    db: "$.service.db",
+    http: {
+      type: "$.interface.http",
+      customResponse: true,
+    },
+  },
+  methods: {
+    getHookType() {
+      // Available hooks: GET https://api.infusionsoft.com/crm/rest/v1/hooks/event_keys
+      throw new Error("Hook type not defined for this source");
+    },
+    getSummary() {
+      throw new Error("Summary defined for this source");
+    },
+    getHookSecretName(): string {
+      return "x-hook-secret";
+    },
+  },
+  hooks: {
+    async activate() {
+      const data: createHookParams = {
+        eventKey: this.getHookType(),
+        hookUrl: this.http.endpoint,
+      };
+
+      const { key }: webhook = await this.infusionsoft.createHook(data);
+
+      this.db.set("hookKey", key);
+    },
+    async deactivate() {
+      const key: string = this.db.get("hookKey");
+
+      await this.infusionsoft.deleteHook({ key });
+    },
+  },
+  async run(data: SourceHttpRunOptions) {
+    const hookSecretName: string = this.getHookSecretName();
+    const hookSecret = data.headers[hookSecretName];
+
+    const httpResponse = {
+      headers: {},
+      status: 200
+    };
+
+    let shouldTriggerEvent = true;
+
+    // If this is a hook verification request:
+    // Do not trigger an event (respond with the secret received)
+    if (hookSecret && data.method === "POST") {
+      shouldTriggerEvent = false;
+      httpResponse.headers[hookSecretName] = hookSecret;
+    };
+
+    this.http.respond(httpResponse);
+
+    // Actual event trigger
+    if (shouldTriggerEvent) {
+      const objectKeys = data.body.object_keys;
+      if (!(objectKeys instanceof Array)) {
+        throw new Error('Unknown data received from Infusionsoft webhook');
+      };
+
+      const promises: Promise<void>[] = objectKeys.map(
+        async (obj: webhookObject) =>
+          new Promise(async (resolve) => {
+            const { apiUrl, id, timestamp } = obj;
+
+            const response = await this.infusionsoft.hookResponseRequest(apiUrl);
+            const data = response.noUrl ? obj : response;
+            const summary = this.getSummary(data);
+
+            this.$emit(data, {
+              id,
+              summary,
+              ts: new Date(timestamp).valueOf(),
+            });
+
+            resolve();
+          })
+      );
+
+      await Promise.allSettled(promises);
+    }
```

Comment:
How about this refactor?
```suggestion
    // If this is a hook verification request:
    // Do not trigger an event (respond with the secret received)
    if (hookSecret && data.method === "POST") {
      return this.http.respond({
        headers: {
          hookSecretName: hookSecret,
        },
        status: 200,
      });
    }

    const { object_keys: objectKeys } = data.body;
    if (!Array.isArray(objectKeys)) {
      throw new Error("Unknown data received from Infusionsoft webhook");
    }

    const promises: Promise<{
      obj: webhookObject;
      response: any;
    }>[] =
      objectKeys.map(async (obj: webhookObject) => ({
        obj,
        response: await this.infusionsoft.hookResponseRequest(obj.apiUrl),
      }));

    const result = await Promise.all(promises);

    result.forEach(({
      obj, response,
    }) => {
      const data = response.noUrl
        ? obj
        : response;
      const summary = this.getSummary(data);
      this.$emit(data, {
        id: obj.id,
        summary,
        ts: new Date(obj.timestamp).valueOf(),
      });
    });
```

Reply:
Great suggestions! The promise wrapping was definitely not ideal and this is much better.

I've `return`'ed when aborting the event emitting, instead of using a variable, as you suggested, however I kept `this.http.respond` as well after that - as the workflow needs to respond with 200 both when it is validating the webhook, and when receiving an event.